src/nbl/asset/interchange/CGLTFLoader.cpp

```diff
@@ -1489,44 +1572,30 @@ namespace nbl
-							// I dont want to have a new shader for every single pipeline just because its skinned
-							// you should create 2x the vertex shaders and register as global builtin in the loader's constructor
 							if (isSkinned)
 							{
-								constexpr std::string_view NBL_SKINNING_OVERRIDE = "#define _NBL_SKINNING_ENABLED_\n";
-
-								auto getOverridenShader = [&](core::smart_refctd_ptr<ICPUSpecializedShader> cpuShader) -> core::smart_refctd_ptr<ICPUSpecializedShader>
-								{
-									const asset::ICPUShader* unspecializedShader = cpuShader->getUnspecialized();
-									assert(unspecializedShader->containsGLSL());
-
-									auto newUnspecializedShader = IGLSLCompiler::createOverridenCopy(unspecializedShader, NBL_SKINNING_OVERRIDE.data(), 69);
-									auto specializedInfo = cpuShader->getSpecializationInfo();
-
-									return core::make_smart_refctd_ptr<asset::ICPUSpecializedShader>(std::move(newUnspecializedShader), std::move(specializedInfo));
-								};
-
-								/*
-									@devshgraphicsprogramming leaving this for you
-								*/
-
-								//! cpuShaders.first = std::move(getOverridenShader(cpuShaders.first));
-								//! cpuShaders.second = std::move(getOverridenShader(cpuShaders.second));
+								if (hasUV) // if both UV and Color defined - we use the UV
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_SKINNED_UV_CACHE_KEY), loadShader(FRAG_SHADER_SKINNED_UV_CACHE_KEY));
+								else if (hasColor)
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_SKINNED_COLOR_CACHE_KEY), loadShader(FRAG_SHADER_SKINNED_COLOR_CACHE_KEY));
+								else
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_SKINNED_NO_UV_COLOR_CACHE_KEY), loadShader(FRAG_SHADER_SKINNED_NO_UV_COLOR_CACHE_KEY));
+							}
+							else
+							{
+								if (hasUV) // if both UV and Color defined - we use the UV
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_UV_CACHE_KEY), loadShader(FRAG_SHADER_UV_CACHE_KEY));
+								else if (hasColor)
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_COLOR_CACHE_KEY), loadShader(FRAG_SHADER_COLOR_CACHE_KEY));
+								else
+									cpuShaders = std::make_pair(loadShader(VERT_SHADER_NO_UV_COLOR_CACHE_KEY), loadShader(FRAG_SHADER_NO_UV_COLOR_CACHE_KEY));
 							}
 
 							return cpuShaders;
 						};
 
 						core::smart_refctd_ptr<ICPURenderpassIndependentPipeline> cpuPipeline;
-						const std::string& pipelineCacheKey = getPipelineCacheKey(primitiveTopology, vertexInputParams); // TODO: add skinning boolean to the pipeline cache key
+						const std::string& pipelineCacheKey = getPipelineCacheKey(primitiveTopology, vertexInputParams, skinningEnabled);
 						{
 							const asset::IAsset::E_TYPE types[]{ asset::IAsset::ET_RENDERPASS_INDEPENDENT_PIPELINE, (asset::IAsset::E_TYPE)0u };
 							auto pipeline_bundle = _override->findCachedAsset(pipelineCacheKey, types, context.loadContext, _hierarchyLevel + ICPUMesh::PIPELINE_HIERARCHYLEVELS_BELOW);
```

Comment:
ignored TODO

Reply:
`skinningEnabled` added to pipeline cache key

Reply:
done myself.

Reply:
btw why skinning influences cache key? some extra bindings?